app/src/main/java/tech/ula/ServerService.kt

```diff
@@ -80,17 +85,100 @@ class ServerService : Service() {
             }
         }
 
-        return Service.START_STICKY
+        val action = intent?.action
+        when (action) {
+            "autostart" -> {
+                autoStartSessions()
+            }
+            "exit" -> {
+                stopServiceAndSessions()
+            }
+            "wakelock" -> {
+                acquireWakeLocks()
+            }
+            "wakerelease" -> {
+                releaseWakeLocks()
+            }
+        }
+        return START_STICKY
+    }
+
+    /**
+     * Fetch all sessions marked to start on boot and start them if they are not already active.
+     */
+    private fun autoStartSessions() {
+        CoroutineScope(Dispatchers.Default).launch {
+            val sessions = UlaDatabase.getInstance(this@ServerService).sessionDao().getSessionsAutoStart()
+            var startedAnySession = false
+            for (session in sessions)
+                if (!session.active) {
+                    startSession(session, false)
+                    startedAnySession = true
+                }
+            if (startedAnySession)
+                acquireWakeLocks()
+        }
+    }
+
+    fun acquireWakeLocks() {
+        // Keep the CPU awake.
+        val powerManager = applicationContext.getSystemService(Context.POWER_SERVICE) as PowerManager
+        wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "ula:Service")
+        wakeLock?.acquire()
+        // Keep the wifi awake.
+        val wm = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
+        wifiLock = wm.createWifiLock(WifiManager.WIFI_MODE_FULL_HIGH_PERF, "ula:Service")
+        wifiLock?.acquire()
+        notificationManager.updateNotification(true)
     }
 
-    // Used in conjunction with manifest attribute `android:stopWithTask="true"`
-    // to clean up when app is swiped away.
-    override fun onTaskRemoved(rootIntent: Intent?) {
-        super.onTaskRemoved(rootIntent)
+    fun releaseWakeLocks() {
+        try {
+            wakeLock?.release()
+            wifiLock?.release()
+        } catch (ignore: Exception) {
+        }
+        wakeLock = null
+        wifiLock = null
+        notificationManager.updateNotification(false)
+    }
+
+    fun stopServiceAndSessions() {
+        releaseWakeLocks()
+        for (session in activeSessions)
+            killSession(session.value)
         stopForeground(true)
         stopSelf()
     }
 
+    /**
+     * Used in conjunction with manifest attribute `android:stopWithTask="true"`
+     * to clean up when app is swiped away.
+     *
+     * Warning : stopWithTask on the manifest is now false because if true, this method is not called
```

Comment:
You can remove the unnecessary parts of this comment.

Reply:
Will do